### ISSUE TITLE
The MPI_Comm_dup() function was called before MPI_INIT was invoked

### ISSUE THREAD

**Issue by deyanmihaylov** [h5py/h5py#2230]
To assist reproducing bugs, please include the following:
 * Operating System: MacOS Ventura 13.0.1
 * Python version: 3.7, 3.8
 * Where Python was acquired: `brew` on mac
 * h5py version: 3.8.0
 * HDF5 version: 1.12.2

I am trying to run the following code (`h5py_write.py`)
```
from mpi4py import MPI
import h5py

rank = MPI.COMM_WORLD.rank  # The process ID (integer 0-3 for 4-process run)

f = h5py.File('parallel_test.hdf5', 'w', driver='mpio', comm=MPI.COMM_WORLD)

dset = f.create_dataset('test', (4,), dtype='i')
dset[rank] = rank

f.close()
```

The output of running
`mpiexec -n 4 python h5py_write.py`
is the following
```
*** The MPI_Comm_dup() function was called before MPI_INIT was invoked.
*** This is disallowed by the MPI standard.
*** Your MPI job will now abort.
[Deyans-MacBook-Pro-Home.local:19927] Local abort before MPI_INIT completed completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
*** The MPI_Comm_dup() function was called before MPI_INIT was invoked.
*** This is disallowed by the MPI standard.
*** Your MPI job will now abort.
[Deyans-MacBook-Pro-Home.local:19930] Local abort before MPI_INIT completed completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
*** The MPI_Comm_dup() function was called before MPI_INIT was invoked.
*** This is disallowed by the MPI standard.
*** Your MPI job will now abort.
[Deyans-MacBook-Pro-Home.local:19928] Local abort before MPI_INIT completed completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
*** The MPI_Comm_dup() function was called before MPI_INIT was invoked.
*** This is disallowed by the MPI standard.
*** Your MPI job will now abort.
[Deyans-MacBook-Pro-Home.local:19929] Local abort before MPI_INIT completed completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
```

I experimented with downgrading versions of `mpi4py` and `h5py` but it gave no effect on the above issue.

**Comment by ajelenak** [comment]
Is the libhdf5 library you are using built for parallel processing? https://docs.h5py.org/en/stable/mpi.html#building-against-parallel-hdf5.

**Comment by deyanmihaylov** [comment]
Yes, it was built with the correct flags for parallel processing.

**Comment by takluyver** [comment]
Is it possible that the [`MPI4PY_RC_INITIALIZE`](https://mpi4py.readthedocs.io/en/latest/mpi4py.html#envvar-MPI4PY_RC_INITIALIZE) environment variable is set?

**Comment by deyanmihaylov** [comment]
There is no output of `echo $MPI4PY_RC_INITIALIZE` both inside and outside of my `python` environment.

**Comment by ndattani** [comment]
@deyanmihaylov  I guess that means that `MPI4PY_RC_INITIALIZE` was not set? Do you get the same error messages when you try the same thing with that environment variable set appropriately?

**Comment by takluyver** [comment]
To be clear, I think it should work correctly without that environment variable set - the default in mpi4py is to initialise MPI when you import `mpi4py.MPI`, which you do in the first line. I just wanted to check that it hadn't somehow got set and overridden the default.

That said, I don't have a better idea for what to try.

**Comment by ndattani** [comment]
@takluyver The output of `echo $MPI4PY_RC_INITIALIZE` was empty though. Perhaps if the echo were to be performed after `from mpi4py import MPI` in the Python script, it would be more clear whether or not the environment variable is set correctly. 

As a Linux user I won't be able to replicate @deyanmihaylov's experience in the MacOS environment. Do you or @ajelenak have access to a machine with MacOS?

**Comment by takluyver** [comment]
I don't have access to a MacOS machine either.

**Comment by derobins** [comment]
MPI should be initialized before you do anything in HDF5. The library will cache the MPI communicator and info when files are open, fapls are copied, etc.

**Comment by ndattani** [comment]
@derobins isn't it initialized though? @takluyver pointed out:
> "the default in mpi4py is to initialise MPI when you import `mpi4py.MPI`"
Since `from mpi4py import MPI` was called in the first line of the OP's script, would MPI not be initialized? Or should does it need `import mpi4py`? I guess if @deyanmihaylov were to tell us the output of `echo $MPI4PY_RC_INITIALIZE` after the  `from mpi4py import MPI` line it would tell us more, but we haven't heard from him since at lest 6 messages ago. 

Perhaps we need someone with MacOS to be able to test if his problem can even be reproduced. You wouldn't happen to have a MacOS machine @derobins?

**Comment by ajelenak** [comment]
I just saw I was mentioned previously here... I do have a macOS laptop but have never used MPI.

**Comment by deyanmihaylov** [comment]
I finally had the change to test this. If I try to output the `MPI4PY_RC_INITIALIZE` environment variable after import MPI, it is defined as `True`. Which makes sense, since scripts which use only `mpi4py` but not `h5py` run fine, it is only when I involve `h5py` that I get this error.

**Comment by ndattani** [comment]
Based on the last comment, which was from the OP, it seems that this is still an open problem.

**Comment by AnubhavdStnfrd** [comment]
Did anyone figure out how to resolve this? Same problem here on macOS Sonoma 14.5, h5py 3.11
